[SPARK-9026] Refactor SimpleFutureAction.onComplete to not launch separate thread for every callback

JoshRosen opened the pull request:
This commit refactors SimpleFutureAction so that registering a callback with `onComplete` does not immediately tie up a thread in the provided execution context.

As @zsxwing noticed in https://github.com/apache/spark/pull/7276#issuecomment-121097747, the existing implementation of SimpleFutureAction.onComplete creates a separate thread to wrap the blocking `awaitResult()` callback:

``` scala
  override def onComplete[U](func: (Try[T]) => U)(implicit executor: ExecutionContext) {
    executor.execute(new Runnable {
      override def run() {
        func(awaitResult())
      }
    })
  }
```

This PR addresses this issue by making JobWaiter into a Future using that future in the implementation of SimpleFutureAction's future methods.

This patch was pair-programmed with @tdas.

JoshRosen committed:
Refactor SimpleFutureAction to not block threads for every onComplete callback.

JoshRosen commented:
@zsxwing, TD and I think that this patch's fix may eliminate the need to introduce a new submitAsyncJob API.  I'm going to push another commit in a few minutes to add some code comments plus more tests, but I just wanted to submit this PR now to get your initial feedback on the approach.

JoshRosen commented on core/src/main/scala/org/apache/spark/scheduler/JobWaiter.scala:
```diff
@@ -28,12 +31,18 @@ private[spark] class JobWaiter[T](
   @volatile
   private var _jobFinished = totalTasks == 0
 
+  if (_jobFinished) {
```
@zsxwing, this `if` statement fixes a subtle bug that we found in your JobWaiter future: if a job has no tasks, then it is marked as finished immediately and `taskSucceeded` will not be called, so we need to complete the promise here.  We noticed this because a test in AsyncRDDActionsSuite was hanging.

JoshRosen commented on core/src/test/resources/log4j.properties:
```diff
@@ -16,7 +16,11 @@
 
 # Set everything to be logged to the file target/unit-tests.log
-log4j.rootCategory=INFO, file
+log4j.rootCategory=INFO, console
```
Whoops, meant to leave this out.  Will revert.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37179 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37179/consoleFull) for   PR 7385 at commit [`11c7450`](https://github.com/apache/spark/commit/11c745095bbaa382a15e6ad1ee1afd2e933e4da8).

JoshRosen committed:
Add some comments

JoshRosen committed:
Back out log4j.properties changes.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37184 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37184/consoleFull) for   PR 7385 at commit [`7b25e6b`](https://github.com/apache/spark/commit/7b25e6b40cf5e193cca16139c3a84775aa1d0927).

JoshRosen committed:
Add regression test

JoshRosen commented:
@tdas, PTAL @ the regression test that I added.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37188 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37188/consoleFull) for   PR 7385 at commit [`d779af8`](https://github.com/apache/spark/commit/d779af8d2fd6b8c64f0e914ee59eb8700d60f8e8).

SparkQA commented:
[Test build #37179 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37179/console) for   PR 7385 at commit [`11c7450`](https://github.com/apache/spark/commit/11c745095bbaa382a15e6ad1ee1afd2e933e4da8).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
[Test build #37184 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37184/console) for   PR 7385 at commit [`7b25e6b`](https://github.com/apache/spark/commit/7b25e6b40cf5e193cca16139c3a84775aa1d0927).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

JoshRosen commented:
Ah, interesting: it looks like this failed some of the JavaAsyncRDDActions tests. I'll investigate.

SparkQA commented:
[Test build #37188 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37188/console) for   PR 7385 at commit [`d779af8`](https://github.com/apache/spark/commit/d779af8d2fd6b8c64f0e914ee59eb8700d60f8e8).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class Least(children: Expression*) extends Expression `
  * `case class Greatest(children: Expression*) extends Expression `

AmplabJenkins commented:
Merged build finished. Test FAILed.

JoshRosen commented on core/src/main/scala/org/apache/spark/FutureAction.scala:
```diff
@@ -116,57 +121,32 @@ class SimpleFutureAction[T] private[spark](jobWaiter: JobWaiter[_], resultFunc:
+  override def onComplete[U](func: (Try[T]) => U)(implicit executor: ExecutionContext): Unit = {
+    jobWaiter.toFuture.onComplete { (jobWaiterResult: Try[Unit]) =>
+      // If the job succeeded, then evaluate the result function; otherwise, preserve the exception.
+      _value = jobWaiterResult.map(_ => resultFunc)
```
@tdas, I think there's a bug here because we'll re-assign to `_value` if there are multiple `onCompletes`.  There's also a race in allowing `_value` to be assigned here, since there's a lag between when the `jobWaiter` future completes and when this callback runs.  Fixing this now...

JoshRosen committed:
Fix race.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37200 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37200/consoleFull) for   PR 7385 at commit [`1e2db7f`](https://github.com/apache/spark/commit/1e2db7f54952aa9338acea6e2b64ca24d61ed2ab).

zsxwing commented:
@JoshRosen thank you for fixing this. Actually, I have another requirement for `submitJob`. Current `submitJob` signature is 
```
  def submitJob[T, U, R](
      rdd: RDD[T],
      processPartition: Iterator[T] => U,
      partitions: Seq[Int],
      resultHandler: (Int, U) => Unit,
      resultFunc: => R): SimpleFutureAction[R]
```
I cannot access `TaskContext` in this method. Could you change `processPartition: Iterator[T] => U` to `processPartition: (TaskContext, Iterator[T]) => U`, or add a new method for it?

JoshRosen commented:
@zsxwing will `TaskContext.get()` work for you?  AFAIK we're trying to avoid the introduction of new `withContext` methods when `TaskContext.get()` will suffice (see some of the old discussions regarding `mapPartitionsWithContext`, etc.).

zsxwing commented:
> @zsxwing will TaskContext.get() work for you? AFAIK we're trying to avoid the introduction of new withContext methods when TaskContext.get() will suffice (see some of the old discussions regarding mapPartitionsWithContext, etc.).

Forgot it. It should work. Thanks.

zsxwing commented on core/src/main/scala/org/apache/spark/FutureAction.scala:
```diff
@@ -116,37 +120,18 @@ class SimpleFutureAction[T] private[spark](jobWaiter: JobWaiter[_], resultFunc:
-      }
-    })
+  override def onComplete[U](func: (Try[T]) => U)(implicit executor: ExecutionContext): Unit = {
+    jobWaiterFuture.map { _ => resultFunc }.onComplete(func)
```
How about adding `private lazy val _resultFunc = resultFunc` and use `resultFunc` in this class? This should avoid calling `resultFunc` multiple times.

JoshRosen replied:
I think that @tdas and I considered this and ended up not doing it because we thought that `resultFunc` would only be computed once, but I suppose it doesn't hurt to be more explicit here.  Even if the current code works, if it's confusing enough to merit a comment then I think we should just be explicit and use a lazy val. I'll update this now to do this.

SparkQA commented:
[Test build #37200 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37200/console) for   PR 7385 at commit [`1e2db7f`](https://github.com/apache/spark/commit/1e2db7f54952aa9338acea6e2b64ca24d61ed2ab).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

JoshRosen committed:
Use lazy val to make ot clear that resultfunc should only be evaluated once

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37245 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37245/consoleFull) for   PR 7385 at commit [`1346313`](https://github.com/apache/spark/commit/1346313c38d8009d849399b654913b95b3646ff5).

JoshRosen commented:
@zsxwing I've updated this to use a `lazy val`; please take another look.

SparkQA commented:
[Test build #37245 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37245/console) for   PR 7385 at commit [`1346313`](https://github.com/apache/spark/commit/1346313c38d8009d849399b654913b95b3646ff5).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

tdas commented on core/src/main/scala/org/apache/spark/FutureAction.scala:
```diff
@@ -154,17 +143,13 @@ class SimpleFutureAction[T] private[spark](jobWaiter: JobWaiter[_], resultFunc:
-      case JobSucceeded => scala.util.Success(resultFunc)
-      case JobFailed(e: Exception) => scala.util.Failure(e)
+    } else {
+      jobWaiter.awaitResult() match {
```
This part seems like a bad hack to use `awaitResult` to get the result. Rather, there should be a `JobWaiter.jobResult` (make it public), that return `Option[JobResult]` and use that.

tdas commented on core/src/main/scala/org/apache/spark/scheduler/JobWaiter.scala:
```diff
@@ -28,12 +31,18 @@ private[spark] class JobWaiter[T](
     resultHandler: (Int, T) => Unit)
   extends JobListener {
 
+  private val promise = Promise[Unit]
```
Stepping back. If we are using promise any ways, why do we need a separate variable called "jobFinished"? The promise is sufficient in keeping the state of whether the job has finished or not.

JoshRosen committed:
Convert JobWaiter into a Future

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37319 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37319/consoleFull) for   PR 7385 at commit [`e08623a`](https://github.com/apache/spark/commit/e08623a5c10d3d0201ab0b4d5b5f2fcbde85bfe7).

JoshRosen commented:
I just pushed a commit which changes JobWaiter to extend `Future[Unit]`, which I think should resolve the most recent round of review feedback.

SparkQA commented:
[Test build #37319 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37319/console) for   PR 7385 at commit [`e08623a`](https://github.com/apache/spark/commit/e08623a5c10d3d0201ab0b4d5b5f2fcbde85bfe7).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

zsxwing commented on core/src/main/scala/org/apache/spark/scheduler/JobWaiter.scala:
```diff
@@ -50,28 +72,19 @@ private[spark] class JobWaiter[T](
-      _jobFinished = true
-      jobResult = JobSucceeded
+      promise.trySuccess()
       this.notifyAll()
```
This line can be removed. Right?

zsxwing commented on core/src/main/scala/org/apache/spark/scheduler/JobWaiter.scala:
```diff
@@ -50,28 +72,19 @@ private[spark] class JobWaiter[T](
-    _jobFinished = true
-    jobResult = JobFailed(exception)
+    promise.tryFailure(exception)
     this.notifyAll()
```
This line can be removed too

zsxwing commented:
LGTM except the minor comments.

JoshRosen committed:
Remove unnecessary notifyAll() cals.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37370 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37370/consoleFull) for   PR 7385 at commit [`b504384`](https://github.com/apache/spark/commit/b504384eb18169190e6322faaf48ad1a05cf07de).

zsxwing commented:
LGTM

SparkQA commented:
[Test build #37370 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37370/console) for   PR 7385 at commit [`b504384`](https://github.com/apache/spark/commit/b504384eb18169190e6322faaf48ad1a05cf07de).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

JoshRosen commented:
@tdas, I'll leave this open if you want to take a final look.  Feel free to merge in order to unblock @zsxwing's other patch.

tdas commented:
@rxin quick look please.

tdas commented on core/src/test/scala/org/apache/spark/FutureActionSuite.scala:
```diff
@@ -49,4 +50,20 @@ class FutureActionSuite
+    try {
+      for (_ <- 1 to 10) {
+        job.onComplete(_ => ())(executionContext)
+        assert(pool.getLargestPoolSize < 10)
```
This looks flaky. Even they are non blocking, there is NO guarantee that one of the 10 scheduled function `_ => ()` will finish by the end of this loop. So it may happen that in the 10th iteration, the previous 9 scheduled function are still not finished, the 10th on gets scheduled, and therefore the pool size = 10.

JoshRosen replied:
I think that my intention when writing this test was to have a test that demonstrated the eagerly-create-a-thread-per-callback problem with the old implementation of SimpleFutureAction.

I don't think that this is flaky but I also don't think that this tests adds much value since we're unlikely to ever switch back to the old inefficient implementation.  I'll just drop this test, since I don't think it's adding any real value right now.

JoshRosen committed:
Remove unnecessary test

JoshRosen committed:
Use success instead of trySuccess; etc.

JoshRosen committed:
Merge remote-tracking branch 'josh/simplefutureaction-refactoring' into simplefutureaction-refactoring

JoshRosen commented:
Ping @rxin, @tdas said that you should probably take a look at this, since you may be able to comment on why JobWaiter wasn't a Future to begin with.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37442 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37442/consoleFull) for   PR 7385 at commit [`c6fdc21`](https://github.com/apache/spark/commit/c6fdc2169f5bb8802b7b2d0019433de8bb0cae66).

rxin commented on core/src/main/scala/org/apache/spark/FutureAction.scala:
```diff
@@ -116,55 +123,27 @@ class SimpleFutureAction[T] private[spark](jobWaiter: JobWaiter[_], resultFunc:
-      case scala.util.Success(res) => res
-      case scala.util.Failure(e) => throw e
-    }
+    jobWaiter.result(atMost)(permit) // Throws exception if the job failed.
```
what thread does this use? does it use some implicit thread pool?

JoshRosen replied:
Let's see... jobWaiter.result delegates to the job waiter's promise.future.result.

Here, I think we have an instance of DefaultPromise, whose `result` (https://github.com/scala/scala/blob/v2.10.4/src/library/scala/concurrent/impl/Promise.scala#L222) is implemented in terms of `ready` (https://github.com/scala/scala/blob/v2.10.4/src/library/scala/concurrent/impl/Promise.scala#L218), which in turn is implemented using the internal `tryAwait` method: https://github.com/scala/scala/blob/v2.10.4/src/library/scala/concurrent/impl/Promise.scala#L194

It looks like this is implemented by scheduling an `onComplete` callback which updates a latch.  This callback runs on `Future`'s InternalCallbackExecutor: https://github.com/scala/scala/blob/v2.10.4/src/library/scala/concurrent/Future.scala#L590

SparkQA commented:
**[Test build #37442 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37442/console)**     for PR 7385 at commit [`c6fdc21`](https://github.com/apache/spark/commit/c6fdc2169f5bb8802b7b2d0019433de8bb0cae66)     after a configured wait of `175m`.

AmplabJenkins commented:
Merged build finished. Test FAILed.

JoshRosen commented:
Jenkins, retest this please.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37505 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37505/consoleFull) for   PR 7385 at commit [`c6fdc21`](https://github.com/apache/spark/commit/c6fdc2169f5bb8802b7b2d0019433de8bb0cae66).

SparkQA commented:
**[Test build #37505 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37505/console)**     for PR 7385 at commit [`c6fdc21`](https://github.com/apache/spark/commit/c6fdc2169f5bb8802b7b2d0019433de8bb0cae66)     after a configured wait of `175m`.

AmplabJenkins commented:
Merged build finished. Test FAILed.

JoshRosen commented:
Uh oh, looks like there's a problem.  I'll investigate this afternoon.

tdas commented:
Seems unrelated random hang. Trying to run this again.

tdas commented:
Jenkins, test this please.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37537 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37537/consoleFull) for   PR 7385 at commit [`c6fdc21`](https://github.com/apache/spark/commit/c6fdc2169f5bb8802b7b2d0019433de8bb0cae66).

SparkQA commented:
**[Test build #37537 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37537/console)**     for PR 7385 at commit [`c6fdc21`](https://github.com/apache/spark/commit/c6fdc2169f5bb8802b7b2d0019433de8bb0cae66)     after a configured wait of `175m`.

AmplabJenkins commented:
Merged build finished. Test FAILed.

tdas commented:
Okay this seems like  recurring problem. @JoshRosen mind taking a look soon.

JoshRosen commented:
Sure, but I'm kinda swamped with some other stuff now so it might take a day before I get to this.

JoshRosen commented:
Ah, interesting: it looks a race condition in the existing JobWaiter has led to flakiness in a unit test.  Here's the Jenkins failure: https://amplab.cs.berkeley.edu/jenkins/job/Spark-Master-SBT/lastCompletedBuild/AMPLAB_JENKINS_BUILD_PROFILE=hadoop1.0,label=centos/testReport/org.apache.spark.scheduler/DAGSchedulerSuite/misbehaved_resultHandler_should_not_crash_DAGScheduler_and_SparkContext/

> java.lang.UnsupportedOperationException: taskSucceeded() called on a finished JobWaiter was not instance of org.apache.spark.scheduler.DAGSchedulerSuiteDummyException

JoshRosen committed:
Merge remote-tracking branch 'origin/master' into simplefutureaction-refactoring

JoshRosen commented:
It looks like the hanging test is DAGSchedulerSuite's "misbehaved resultHandler should not crash DAGScheduler and SparkContext" test.  Here's the relevant part of the unit tests log:

```
15/07/17 09:05:49.333 pool-1-thread-1-ScalaTest-running-DAGSchedulerSuite INFO SparkContext: Starting job: runJob at DAGSchedulerSuite.scala:762
15/07/17 09:05:49.346 dag-scheduler-event-loop INFO DAGScheduler: Got job 0 (runJob at DAGSchedulerSuite.scala:762) with 1 output partitions (allowLocal=true)
15/07/17 09:05:49.346 dag-scheduler-event-loop INFO DAGScheduler: Final stage: ResultStage 0(runJob at DAGSchedulerSuite.scala:762)
15/07/17 09:05:49.347 dag-scheduler-event-loop INFO DAGScheduler: Parents of final stage: List()
15/07/17 09:05:49.347 dag-scheduler-event-loop INFO DAGScheduler: Missing parents: List()
15/07/17 09:05:49.348 dag-scheduler-event-loop INFO DAGScheduler: Computing the requested partition locally
15/07/17 09:05:49.353 pool-1-thread-1-ScalaTest-running-DAGSchedulerSuite INFO DAGScheduler: Job 0 failed: runJob at DAGSchedulerSuite.scala:762, took 0.019835 s
15/07/17 09:05:49.362 pool-1-thread-1-ScalaTest-running-DAGSchedulerSuite INFO SparkContext: Starting job: runJob at DAGSchedulerSuite.scala:773
15/07/17 09:05:49.363 dag-scheduler-event-loop INFO DAGScheduler: Got job 1 (runJob at DAGSchedulerSuite.scala:773) with 2 output partitions (allowLocal=false)
15/07/17 09:05:49.363 dag-scheduler-event-loop INFO DAGScheduler: Final stage: ResultStage 1(runJob at DAGSchedulerSuite.scala:773)
15/07/17 09:05:49.363 dag-scheduler-event-loop INFO DAGScheduler: Parents of final stage: List()
15/07/17 09:05:49.363 dag-scheduler-event-loop INFO DAGScheduler: Missing parents: List()
15/07/17 09:05:49.366 dag-scheduler-event-loop INFO DAGScheduler: Submitting ResultStage 1 (ParallelCollectionRDD[1] at parallelize at DAGSchedulerSuite.scala:772), which has no missing parents
15/07/17 09:05:49.452 dag-scheduler-event-loop INFO MemoryStore: ensureFreeSpace(1048) called with curMem=0, maxMem=1691332116
15/07/17 09:05:49.453 dag-scheduler-event-loop INFO MemoryStore: Block broadcast_0 stored as values in memory (estimated size 1048.0 B, free 1613.0 MB)
15/07/17 09:05:49.459 dag-scheduler-event-loop INFO MemoryStore: ensureFreeSpace(769) called with curMem=1048, maxMem=1691332116
15/07/17 09:05:49.459 dag-scheduler-event-loop INFO MemoryStore: Block broadcast_0_piece0 stored as bytes in memory (estimated size 769.0 B, free 1613.0 MB)
15/07/17 09:05:49.461 sparkDriver-akka.actor.default-dispatcher-3 INFO BlockManagerInfo: Added broadcast_0_piece0 in memory on localhost:61512 (size: 769.0 B, free: 1613.0 MB)
15/07/17 09:05:49.463 dag-scheduler-event-loop INFO SparkContext: Created broadcast 0 from broadcast at DAGScheduler.scala:901
15/07/17 09:05:49.465 dag-scheduler-event-loop INFO DAGScheduler: Submitting 2 missing tasks from ResultStage 1 (ParallelCollectionRDD[1] at parallelize at DAGSchedulerSuite.scala:772)
15/07/17 09:05:49.466 dag-scheduler-event-loop INFO TaskSchedulerImpl: Adding task set 1.0 with 2 tasks
15/07/17 09:05:49.488 sparkDriver-akka.actor.default-dispatcher-3 INFO TaskSetManager: Starting task 0.0 in stage 1.0 (TID 0, localhost, PROCESS_LOCAL, 1369 bytes)
15/07/17 09:05:49.491 Executor task launch worker-0 INFO Executor: Running task 0.0 in stage 1.0 (TID 0)
15/07/17 09:05:49.504 Executor task launch worker-0 INFO Executor: Finished task 0.0 in stage 1.0 (TID 0). 810 bytes result sent to driver
15/07/17 09:05:49.506 sparkDriver-akka.actor.default-dispatcher-3 INFO TaskSetManager: Starting task 1.0 in stage 1.0 (TID 1, localhost, PROCESS_LOCAL, 1426 bytes)
15/07/17 09:05:49.506 Executor task launch worker-0 INFO Executor: Running task 1.0 in stage 1.0 (TID 1)
15/07/17 09:05:49.507 task-result-getter-0 INFO TaskSetManager: Finished task 0.0 in stage 1.0 (TID 0) in 30 ms on localhost (1/2)
15/07/17 09:05:49.508 pool-1-thread-1-ScalaTest-running-DAGSchedulerSuite INFO DAGScheduler: Job 1 failed: runJob at DAGSchedulerSuite.scala:773, took 0.145836 s
15/07/17 09:05:49.509 Executor task launch worker-0 INFO Executor: Finished task 1.0 in stage 1.0 (TID 1). 810 bytes result sent to driver
15/07/17 09:05:49.511 task-result-getter-1 INFO TaskSetManager: Finished task 1.0 in stage 1.0 (TID 1) in 6 ms on localhost (2/2)
15/07/17 09:05:49.512 dag-scheduler-event-loop INFO DAGScheduler: ResultStage 1 (runJob at DAGSchedulerSuite.scala:773) finished in 0.041 s
15/07/17 09:05:49.512 task-result-getter-1 INFO TaskSchedulerImpl: Removed TaskSet 1.0, whose tasks have all completed, from pool
15/07/17 09:05:49.516 dag-scheduler-event-loop ERROR DAGSchedulerEventProcessLoop: DAGSchedulerEventProcessLoop failed; shutting down SparkContext
java.lang.IllegalStateException: Promise already completed.
	at scala.concurrent.Promise$class.complete(Promise.scala:55)
	at scala.concurrent.impl.Promise$DefaultPromise.complete(Promise.scala:153)
	at scala.concurrent.Promise$class.failure(Promise.scala:107)
	at scala.concurrent.impl.Promise$DefaultPromise.failure(Promise.scala:153)
	at org.apache.spark.scheduler.JobWaiter.jobFailed(JobWaiter.scala:86)
	at org.apache.spark.scheduler.DAGScheduler.handleTaskCompletion(DAGScheduler.scala:1059)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:1508)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:1472)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:1461)
	at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:48)
15/07/17 09:05:49.516 pool-1-thread-1-ScalaTest-running-DAGSchedulerSuite INFO SparkContext: Starting job: count at DAGSchedulerSuite.scala:783
15/07/17 09:05:49.519 dag-scheduler-event-loop INFO DAGScheduler: Stopping DAGScheduler
15/07/17 09:08:48.579 sparkDriver-akka.actor.default-dispatcher-4 WARN HeartbeatReceiver: Removing executor driver with no recent heartbeats: 179561 ms exceeds timeout 120000 ms
15/07/17 09:08:48.580 sparkDriver-akka.actor.default-dispatcher-4 ERROR TaskSchedulerImpl: Lost executor driver on localhost: Executor heartbeat timed out after 179561 ms
```

It looks like the problem is that `jobFailed` can be called multiple times if a task that runs in `DAGScheduler.runLocallyWithinThread` fails.  We get one call from `runLocallyWithinThread` and another call inside of `DAGScheduler.handleTaskCompletion`.

I guess this is actually a subtlety that was overlooked in the old code.  I guess one approach is to make `jobFailed` into an idempotent method.  Let me try that now...

JoshRosen committed:
Make jobFailed idempotent

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37647 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37647/consoleFull) for   PR 7385 at commit [`1a19268`](https://github.com/apache/spark/commit/1a19268fda972f6103bcd50f376c0f57a238b394).

SparkQA commented:
[Test build #37647 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37647/console) for   PR 7385 at commit [`1a19268`](https://github.com/apache/spark/commit/1a19268fda972f6103bcd50f376c0f57a238b394).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `abstract class UnsafeProjection extends Projection `
  * `case class FromUnsafeProjection(fields: Seq[DataType]) extends Projection `
  * `abstract class BaseProjection extends Projection `
  * `    class SpecificProjection extends $`
  * `    class SpecificProjection extends $`

AmplabJenkins commented:
Merged build finished. Test FAILed.

JoshRosen commented:
Jenkins, retest this please.

JoshRosen commented:
This latest failure seems to be due to a DistributedSuite test that may have recently become flaky.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37654 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37654/consoleFull) for   PR 7385 at commit [`1a19268`](https://github.com/apache/spark/commit/1a19268fda972f6103bcd50f376c0f57a238b394).

SparkQA commented:
[Test build #37654 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37654/console) for   PR 7385 at commit [`1a19268`](https://github.com/apache/spark/commit/1a19268fda972f6103bcd50f376c0f57a238b394).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `abstract class UnsafeProjection extends Projection `
  * `case class FromUnsafeProjection(fields: Seq[DataType]) extends Projection `
  * `abstract class BaseProjection extends Projection `
  * `    class SpecificProjection extends $`
  * `    class SpecificProjection extends $`

AmplabJenkins commented:
Merged build finished. Test PASSed.

tdas commented:
Jenkins, test this please.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37673 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37673/consoleFull) for   PR 7385 at commit [`1a19268`](https://github.com/apache/spark/commit/1a19268fda972f6103bcd50f376c0f57a238b394).

SparkQA commented:
[Test build #37673 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37673/console) for   PR 7385 at commit [`1a19268`](https://github.com/apache/spark/commit/1a19268fda972f6103bcd50f376c0f57a238b394).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

JoshRosen commented:
Jenkins, retest this please.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37677 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37677/consoleFull) for   PR 7385 at commit [`1a19268`](https://github.com/apache/spark/commit/1a19268fda972f6103bcd50f376c0f57a238b394).

SparkQA commented:
[Test build #37677 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37677/console) for   PR 7385 at commit [`1a19268`](https://github.com/apache/spark/commit/1a19268fda972f6103bcd50f376c0f57a238b394).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

JoshRosen commented:
Jenkins, retest this please.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37706 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37706/consoleFull) for   PR 7385 at commit [`1a19268`](https://github.com/apache/spark/commit/1a19268fda972f6103bcd50f376c0f57a238b394).

SparkQA commented:
[Test build #37706 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37706/console) for   PR 7385 at commit [`1a19268`](https://github.com/apache/spark/commit/1a19268fda972f6103bcd50f376c0f57a238b394).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

JoshRosen commented:
Jenkins, retest this please.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #37745 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37745/consoleFull) for   PR 7385 at commit [`1a19268`](https://github.com/apache/spark/commit/1a19268fda972f6103bcd50f376c0f57a238b394).

SparkQA commented:
[Test build #37745 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/37745/console) for   PR 7385 at commit [`1a19268`](https://github.com/apache/spark/commit/1a19268fda972f6103bcd50f376c0f57a238b394).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

zsxwing commented on core/src/main/scala/org/apache/spark/scheduler/JobWaiter.scala:
```diff
@@ -50,28 +72,23 @@ private[spark] class JobWaiter[T](
+    // There are certain situations where jobFailed can be called multiple times for the same
+    // job. We guard against this by making this method idempotent.
+    if (!isCompleted) {
+      promise.failure(exception)
```
Looks `tryFailure` would be simpler.

zsxwing commented:
retest this please

zsxwing commented:
@tdas could you take a look at this PR again? I think it's better to merge this one at first.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #61 has started](https://amplab.cs.berkeley.edu/jenkins/job/SlowSparkPullRequestBuilder/61/consoleFull) for   PR 7385 at commit [`1a19268`](https://github.com/apache/spark/commit/1a19268fda972f6103bcd50f376c0f57a238b394).

SparkQA commented:
[Test build #61 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SlowSparkPullRequestBuilder/61/console) for   PR 7385 at commit [`1a19268`](https://github.com/apache/spark/commit/1a19268fda972f6103bcd50f376c0f57a238b394).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38083 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38083/consoleFull) for   PR 7385 at commit [`1a19268`](https://github.com/apache/spark/commit/1a19268fda972f6103bcd50f376c0f57a238b394).

SparkQA commented:
[Test build #38083 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38083/console) for   PR 7385 at commit [`1a19268`](https://github.com/apache/spark/commit/1a19268fda972f6103bcd50f376c0f57a238b394).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
[Test build #1162 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1162/consoleFull) for   PR 7385 at commit [`1a19268`](https://github.com/apache/spark/commit/1a19268fda972f6103bcd50f376c0f57a238b394).

SparkQA commented:
[Test build #1162 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1162/console) for   PR 7385 at commit [`1a19268`](https://github.com/apache/spark/commit/1a19268fda972f6103bcd50f376c0f57a238b394).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

JoshRosen committed:
Merge remote-tracking branch 'origin/master' into simplefutureaction-refactoring

JoshRosen commented:
;Jenkins, retest this please

JoshRosen commented:
Jenkins, retest this please.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38119 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38119/consoleFull) for   PR 7385 at commit [`1a19268`](https://github.com/apache/spark/commit/1a19268fda972f6103bcd50f376c0f57a238b394).

SparkQA commented:
[Test build #38119 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38119/console) for   PR 7385 at commit [`1a19268`](https://github.com/apache/spark/commit/1a19268fda972f6103bcd50f376c0f57a238b394).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

JoshRosen committed:
Merge remote-tracking branch 'origin/master' into simplefutureaction-refactoring

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38258 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38258/consoleFull) for   PR 7385 at commit [`692b3a4`](https://github.com/apache/spark/commit/692b3a4db8e94ee69298dbf3c9b343c3694c4954).

SparkQA commented:
[Test build #38258 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38258/console) for   PR 7385 at commit [`692b3a4`](https://github.com/apache/spark/commit/692b3a4db8e94ee69298dbf3c9b343c3694c4954).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

JoshRosen committed:
Exception -> Throwable

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38280 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38280/consoleFull) for   PR 7385 at commit [`17edbcd`](https://github.com/apache/spark/commit/17edbcd06086b6a8cad922b4c535eb2a6265b2e3).

SparkQA commented:
[Test build #38280 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38280/console) for   PR 7385 at commit [`17edbcd`](https://github.com/apache/spark/commit/17edbcd06086b6a8cad922b4c535eb2a6265b2e3).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

SparkQA commented:
[Test build #1191 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1191/consoleFull) for   PR 7385 at commit [`17edbcd`](https://github.com/apache/spark/commit/17edbcd06086b6a8cad922b4c535eb2a6265b2e3).

SparkQA commented:
[Test build #1192 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1192/consoleFull) for   PR 7385 at commit [`17edbcd`](https://github.com/apache/spark/commit/17edbcd06086b6a8cad922b4c535eb2a6265b2e3).

SparkQA commented:
[Test build #1191 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1191/console) for   PR 7385 at commit [`17edbcd`](https://github.com/apache/spark/commit/17edbcd06086b6a8cad922b4c535eb2a6265b2e3).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
[Test build #1192 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1192/console) for   PR 7385 at commit [`17edbcd`](https://github.com/apache/spark/commit/17edbcd06086b6a8cad922b4c535eb2a6265b2e3).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

JoshRosen commented:
Jenkins, retest this please.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38573 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38573/consoleFull) for   PR 7385 at commit [`17edbcd`](https://github.com/apache/spark/commit/17edbcd06086b6a8cad922b4c535eb2a6265b2e3).

SparkQA commented:
[Test build #38573 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38573/console) for   PR 7385 at commit [`17edbcd`](https://github.com/apache/spark/commit/17edbcd06086b6a8cad922b4c535eb2a6265b2e3).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
**[Test build #2093 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2093/consoleFull)** for PR 7385 at commit [`17edbcd`](https://github.com/apache/spark/commit/17edbcd06086b6a8cad922b4c535eb2a6265b2e3).

SparkQA commented:
**[Test build #2093 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2093/consoleFull)** for PR 7385 at commit [`17edbcd`](https://github.com/apache/spark/commit/17edbcd06086b6a8cad922b4c535eb2a6265b2e3).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

zsxwing commented:
@JoshRosen could you close this one, since  #9264 fixes the issue as well? Thanks!

jaceklaskowski commented:
@zsxwing @JoshRosen Does the comment need attention since the pr is closed,  https://github.com/apache/spark/blob/master/streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala#L438?

rxin commented:
That's a good find. 

cc @zsxwing

zsxwing commented:
Thanks @jaceklaskowski I will submit a PR to remove `submitJobThreadPool`
